folium/features.py

```diff
@@ -288,21 +281,63 @@ def render(self, **kwargs):
             </style>
             """).render(this=self, **kwargs)), name=self.get_name())
 
-        figure.header.add_child(
-            JavascriptLink('https://d3js.org/d3.v3.min.js'),
-            name='d3')
+        if vegalite_major_version == '1':
+            self._embed_vegalite_v1(figure)
+        elif vegalite_major_version == '2':
+            self._embed_vegalite_v2(figure)
+        elif vegalite_major_version == '3':
+            self._embed_vegalite_v3(figure)
+        else:
+            # Version 2 is assumed as the default, if no version is given in the schema.
+            self._embed_vegalite_v2(figure)
+
+    def _get_vegalite_major_versions(self, spec):
+        try:
+            schema = spec['$schema']
+        except KeyError:
+            major_version = None
+        else:
+            major_version = schema.split('/')[-1].split('.')[0].lstrip('v')
 
-        figure.header.add_child(
-            JavascriptLink('https://cdnjs.cloudflare.com/ajax/libs/vega/2.6.5/vega.min.js'),  # noqa
-            name='vega')
+        return major_version
 
-        figure.header.add_child(
-            JavascriptLink('https://cdnjs.cloudflare.com/ajax/libs/vega-lite/1.3.1/vega-lite.min.js'),  # noqa
-            name='vega-lite')
+    def _embed_vegalite_v3(self, figure):
+        self._vega_embed()
 
-        figure.header.add_child(
-            JavascriptLink('https://cdnjs.cloudflare.com/ajax/libs/vega-embed/2.2.0/vega-embed.min.js'),  # noqa
-            name='vega-embed')
+        figure.header.add_child(JavascriptLink('https://cdn.jsdelivr.net/npm/vega@4'), name='vega')
+        figure.header.add_child(JavascriptLink('https://cdn.jsdelivr.net/npm/vega-lite@3'), name='vega-lite')
+        figure.header.add_child(JavascriptLink('https://cdn.jsdelivr.net/npm/vega-embed@3'), name='vega-embed')
+
+    def _embed_vegalite_v2(self, figure):
+        self._vega_embed()
+
+        figure.header.add_child(JavascriptLink('https://cdn.jsdelivr.net/npm/vega@3'), name='vega')
+        figure.header.add_child(JavascriptLink('https://cdn.jsdelivr.net/npm/vega-lite@2'), name='vega-lite')
+        figure.header.add_child(JavascriptLink('https://cdn.jsdelivr.net/npm/vega-embed@3'), name='vega-embed')
+
+    def _vega_embed(self):
+        self._parent.script.add_child(Element(Template("""
+                    const spec = {{this.json}};
```

Comment:
The `spec` here will need a unique name or we won't be able to display more than 1 object.

Reply:
I found a simple fix by just avoiding using the `spec` variable. However, we should test against this case for the future. Unfortunately I have no idea how to do this, as the error is happening on the JS side.

Reply:
> I found a simple fix by just avoiding using the `spec` variable.

Great!

> However, we should test against this case for the future. Unfortunately I have no idea how to do this, as the error is happening on the JS side. Just executing a test with two VegaLite elements does not throw an error.
> Putting an example in the example notebooks might be a way, but could look a bit awkward and is also not caught by pytest.

Yeah, that is hard. We can create a map and compare an expected output but that may not catch errors like that. We do build all the maps in the example gallery and check them ~~always~~ (sometimes) before merging a PR.

Reply:
Maybe one could start a headless selenium browser and check if the map and the charts are there. But that would be maybe a bit two much for this PR. I can easily provide a notebook example for now and try to tackle more elaborate testing in a separate PR if that is fine for you.

Reply:
That kind of testing is in our TODO list for a long time and I agree, it is something for another PR.

This one is good to go, just waiting for @Conengmo last :+1: to merge it.